### DISCUSSION TITLE
TE data for Tread workouts still broken in v4.1.0

### DISCUSSION THREAD

**Discussion by kocolosk** [philosowaffle/peloton-to-garmin#651] [TE / TSS / Vo2 / Intensity Minutes]
**Describe the bug**
The fix in #578 to update the `CyclingDevice` manufacturer worked to restore Training Effect data for Bike workouts, but I'm finding that my Tread workouts are still not reporting TE data over the past few weeks.

**To Reproduce**
Do a Tread workout. The device is identified as a Forerunner 945. No Aerobic TE is reported in Garmin Connect.

**Installation (please complete the following information):**
GH Action, using both the `console-stable` and `console-latest` images

**Additional context**
I do see the following in [Garmin's FAQ](https://support.garmin.com/en-US/?faq=lWqSVlq3w76z5WoihLy5f8):

> NOTE: Indoor runs completed using a supported third-party app, such as Zwift, can affect your training load and therefore impact your Training Status; however, indoor runs will not populate a VO2 max. **No activities, including Indoor Runs, uploaded via third-party apps will generate a Training Effect.**

I suppose this is why the default device is a Forerunner watch? Not sure. I do know Training Effect on my Tread runs was reporting correctly as late as 12/14/23.

**Comment by kocolosk** [comment]
Update: copying the updated Tacx device info from #578 as a `DEVICE_INFO` secret fixed the training effect data for my run. A bit odd to see a cycling trainer as the device for a run but I'll take it :) Perhaps that ought to be the default across the board, if there's no more elegant solution?

I also tried grabbing the device info from an exported TCX of one of my outdoor runs with my Fenix 6 and using that, but it did not work.

**Comment by philosowaffle** [comment]
I'm really surprised using the Fenix 6 device did not work but the taxc did.

I don't have a Tread to do my own testing with, but if there is enough general consensus on switching to Tacx for Runs I'm happy to make it the default (as weird as it is).

**Comment by mgsimon** [comment]
I think the VO2 max and training load calculations are done on a Garmin device only and the important point might be **post** processing for the listed application in the Garmin FAQ.

`p2g (Tacx) => connect => Garmin device (post processing) => connect`

**Comment by jayflan** [comment]
Hi all.  Wondering about the Garmin device (post processing) as I'm having an issue not getting a training effect/load for my first peloton bike+ ride.  (Sorry if I should have commented elsewhere)  

Anyone still having any issues?

Besides the TE/Load not showing (calculating), everything else ran smoothly using the Docker WebUI on my MacBook.  

The stats in GC look good and **I can see the workout on my Instinct 2x watch**.  Also shows up as the Tacx Windows App (1.3) on GC.

Yet.......still no TE/Load calc on either the watch or GC.  (Unlike my walks on my watch which do show both TE and Load values)

Anyone have any thoughts?  (Using stable build 4.1.0 build)

**Comment by philosowaffle** [comment]
In thinking about whether to change this default (and all of the recent TE/TSS/V02 related questions), I've gone down a bit of a rabbit hole in re-thinking how P2G handles Garmin Device Info... namely that it is not very user friendly and it does not provide enough flexibility to configure at the per Exercise Type level.

So I've been playing around with the below.  My thinking is this achieves a few things:
1. Increases visibility into what Devices will be used for each workout type
2. Allows configuring a global fallback Device
3. Allows configuring a Device per Exercise Type
4. Provides a few known Devices (tacx, epic, forerunner) for easier configuration and experimentation

**This change is not available yet. I just got the MVP working, it needs a lot more testing before I can hand it off.  Early preview in case there are any strong negative reactions to this approach.**

### UI Screenshots
(click these to see larger version)
![2024-01-21_10-20-55](https://github.com/philosowaffle/peloton-to-garmin/assets/7275041/2dfe7a7f-e831-412c-b9e3-e87770b78125)

![2024-01-21_10-21-45](https://github.com/philosowaffle/peloton-to-garmin/assets/7275041/1b40f31d-d7bb-47fa-ab43-e0d6e601b8ae)

### JSON Config
(note the new `deviceInfoSettings` object.  `deviceInfoPath` will continue to work, but will be marked as deprecated)
```json
"format": {
      "fit": true,
      "json": false,
      "tcx": false,
      "saveLocalCopy": false,
      "includeTimeInHRZones": false,
      "includeTimeInPowerZones": false,
      "deviceInfoPath": "",
      "deviceInfoSettings": {
        "none": {
          "name": "Forerunner 945",
          "unitId": 1,
          "productID": 3113,
          "manufacturerId": 1,
          "version": {
            "versionMajor": 19,
            "versionMinor": 2.0,
            "buildMajor": 0,
            "buildMinor": 0.0
          }
        },
        "cycling": {
          "name": "TacxTrainingAppWin",
          "unitId": 1,
          "productID": 20533,
          "manufacturerId": 89,
          "version": {
            "versionMajor": 1,
            "versionMinor": 30.0,
            "buildMajor": 0,
            "buildMinor": 0.0
          }
        },
        "rowing": {
          "name": "Epix",
          "unitId": 3413684246,
          "productID": 3943,
          "manufacturerId": 1,
          "version": {
            "versionMajor": 10,
            "versionMinor": 43.0,
            "buildMajor": 0,
            "buildMinor": 0.0
          }
        },
        "circuit": {
          "name": "ads",
          "unitId": 2,
          "productID": 0,
          "manufacturerId": 1,
          "version": {
            "versionMajor": 0,
            "versionMinor": 0.0,
            "buildMajor": 0,
            "buildMinor": 0.0
          }
        },
        "strength": {
          "name": "Forerunner 945",
          "unitId": 1,
          "productID": 3113,
          "manufacturerId": 1,
          "version": {
            "versionMajor": 19,
            "versionMinor": 2.0,
            "buildMajor": 0,
            "buildMinor": 0.0
          }
        }
      },
      "workoutTitleTemplate": "{{PelotonWorkoutTitle}}{{#if PelotonInstructorName}} with {{PelotonInstructorName}}{{/if}}",
      "cycling": {
        "preferredLapType": 0
      },
      "running": {
        "preferredLapType": 0
      },
      "rowing": {
        "preferredLapType": 0
      },
      "strength": {
        "defaultSecondsPerRep": 3
      }
    },
```

**Comment by philosowaffle** [comment]
- The new way to customize Device Info is available now on the `latest` docker tags.  
- The preview Windows exe can be downloaded [here](https://github.com/philosowaffle/peloton-to-garmin/actions/runs/7775791959/artifacts/1218656402).  
- The documentation is [here](https://philosowaffle.github.io/peloton-to-garmin/master/configuration/format/#customizing-the-garmin-device-associated-with-the-workout).

This change **should be** backwards compatible and gracefully handle both versions of configuring Device Info, but it is recommended to move onto the new standard.

- If you are using the WebUI or Windows Exe, your previous DeviceInfoPath settings will be migrated to the new settings style automatically.
- If you are using a headless version of P2G (like GitHubActions), then you will need to manually switch over to the new settings style.  The values found in the XML file should map one to one with the new JSON config options.

These changes will be included in the next 4.2.0 release.

After the 4.2.0 release is made public, I plan to transfer this issue to the Discussion Forum.  I hope that as people discover Device Configurations that work for TE/TSS/VO2, those can be shared and easily configured via these new configuration options.

**Comment by jayflan** [comment]
Thank you @philosowaffle!

**Comment by jayflan** [comment]
FYI.  Updated to the latest version, synced today's ride, and all looks well!

**Comment by kocolosk** [comment]
Cool. I updated my GH Actions to use the new DeviceInfo format and ran them using the `console-latest` image tag.

As before with the XML config, I found that the default Forerunner 945 is not reporting TE data for Tread workouts, but if I set the `TacxTrainingAppWin` configuration as the `"none"` default in the DeviceInfoSettings then the post-processing kicks in.

Thanks!

**Comment by RicardoNiepel** [comment]
> * The new way to customize Device Info is available now on the `latest` docker tags.
> * The preview Windows exe can be downloaded [here](https://github.com/philosowaffle/peloton-to-garmin/actions/runs/7775791959/artifacts/1218656402).
> * The documentation is [here](https://philosowaffle.github.io/peloton-to-garmin/master/configuration/format/#customizing-the-garmin-device-associated-with-the-workout).
> 
> This change **should be** backwards compatible and gracefully handle both versions of configuring Device Info, but it is recommended to move onto the new standard.
> 
> * If you are using the WebUI or Windows Exe, your previous DeviceInfoPath settings will be migrated to the new settings style automatically.
> * If you are using a headless version of P2G (like GitHubActions), then you will need to manually switch over to the new settings style.  The values found in the XML file should map one to one with the new JSON config options.
> 
> These changes will be included in the next 4.2.0 release.
> 
> After the 4.2.0 release is made public, I plan to transfer this issue to the Discussion Forum. I hope that as people discover Device Configurations that work for TE/TSS/VO2, those can be shared and easily configured via these new configuration options.

@philosowaffle Is there again an known issue? I just used your preview version and TE/TSS/VO2 is once again missing from bike activity on Garmin Connect.

**Comment by philosowaffle** [comment]
[v4.2.0](https://github.com/philosowaffle/peloton-to-garmin/releases/tag/v4.2.0) is released now. Experimenting with different device configurations is currently the best option for keeping up with Garmins TE/TSS/VO2 requirements.

I'm converting this issue over to a Discussion now. I hope as people discover Device Configurations that work for TE/TSS/VO2, those can be shared and easily configured via these new configuration options.
